test_functional.py

```diff
@@ -65,42 +24,51 @@ def client_code():
     return client_conn
 
 
-def test_function(client_code):
-    msg = "do you hear me at all?"
+def test_functional(client_code):
+    msg = b"GET /index.html HTTP/1.1\r\nHost:"
 
     try:
         client_code.sendall(msg)
         output = ""
         while True:
-            part = client_code.recv(16)
+            part = client_code.recv(1024)
             assert part
             output = output + part
-            if len(part) < 16:
+            if len(part) < 1024:
                 client_code.shutdown(socket.SHUT_WR)
                 client_code.close()
                 break
-        assert '<html>' in output
-        assert '200' in output
-        assert 'do you hear me at all?' in output
     except Exception as e:
         print e
+    assert b'200' in output
 
 
-def test_500_function():
+def test_bad_request_functional(client_code):
     msg = "do you hear me at all?"
 
     try:
         client_code.sendall(msg)
-        output = ""
+        output = b""
         while True:
-            part = client_code.recv(16)
+            part = client_code.recv(1024)
             assert part
             output = output + part
-            if len(part) < 16:
+            if len(part) < 1024:
                 client_code.shutdown(socket.SHUT_WR)
                 client_code.close()
                 break
-        assert '<html>' in output
-        assert '500' in output
     except Exception as e:
         print e
+    assert b"400" in output
+
+
+def test_bad_parse():
+    bad_request = b"some_bad_request"
+    with pytest.raises(ValueError):
+        parse_request(bad_request)
+
+
+def test_parse_request():
+    bad_request = b"GET /index.html HTTP/1.1\r\nHost:"
+    response = parse_request(bad_request)
+    assert b"/index.html" in response
```

Comment:
good start.  YOu need to test more about a request to expose the problems with your `parse_request` method.